### ISSUE TITLE
[Bug]: can't use Symfony HttpClient 

### ISSUE THREAD

**Issue by tacman** [ToshY/BunnyNet-PHP#121]
### Package version(s) affected

4.2.1

### Description

I'd like to use the Symfony Client, which implement the psr interface (second on this list: https://packagist.org/providers/psr/http-client-implementation)

### How to reproduce

But when I try to initialize the library
```php
       $bunnyClient = new BunnyClient(
            client: (new HttpClient()) // or inject the default http client, same problem
        );
```

I get ToshY\BunnyNet\Client\BunnyClient::__construct(): Argument #1 ($client) must be of type Psr\Http\Client\ClientInterface, Symfony\Component\HttpClient\TraceableHttpClient given

### Possible Solution

I think the issue is the type, Symfony uses cache contracts. 

use Symfony\Contracts\HttpClient\HttpClientInterface;

### Additional Context

I was able to get it to work by installing 

composer require php-http/discovery

and finding the implementation

```
        $httpClient =  Psr18ClientDiscovery::find();
// Create a BunnyClient using any HTTP client implementing "Psr\Http\Client\ClientInterface".
        $bunnyClient = new BunnyClient(
            client: $httpClient
        );
```

I'm sure I'm just confused about the interfaces, perhaps you could add something to the documentation as to how to use your own http client. 

Tx.

**Comment by ToshY** [comment]
Hey @tacman 👋 

Okay let's just take it step-by-step then.

> I'd like to use the Symfony Client, which implement the psr interface (second on this list: https://packagist.org/providers/psr/http-client-implementation)

While Symfony provides a PSR-18 implementation, it does not automatically mean the default `HttpClient` is also PSR-18 compatible.

> But when I try to initialize the library
> 
>        $bunnyClient = new BunnyClient(
>             client: (new HttpClient()) // or inject the default http client, same problem
>         );
> I get ToshY\BunnyNet\Client\BunnyClient::__construct(): Argument https://github.com/ToshY/BunnyNet-PHP/pull/1 ($client) must be of type Psr\Http\Client\ClientInterface, Symfony\Component\HttpClient\TraceableHttpClient given

As you've noticed, the exception shows that a [`Symfony\Component\HttpClient\TraceableHttpClient`](https://github.com/symfony/http-client/blob/7.0/TraceableHttpClient.php) was given, which does **not** implement the `Psr\Http\Client\ClientInterface` but it's own `Symfony\Contracts\HttpClient\HttpClientInterface`. The same goes for the majority of other Symfony `HttpClient` implementations.

The solution however can be found in the [HTTP Client documentation (section "PSR-18 and PSR-17")](https://symfony.com/doc/current/http_client.html#psr-18-and-psr-17).

> I'm sure I'm just confused about the interfaces, perhaps you could add something to the documentation as to how to use your own http client.

The current documentation states with a comment that the HTTP client you want to use must implement `Psr\Http\Client\ClientInterface`, and also gives as an example to use `new \Symfony\Component\HttpClient\Psr18Client()`:

```php
// Create a BunnyClient using any HTTP client implementing "Psr\Http\Client\ClientInterface".
$bunnyClient = new BunnyClient(
    client: new \Symfony\Component\HttpClient\Psr18Client(),
);
```

From my perspective I'm not sure what's unclear about this, but if you have suggestions on how you think this can be improved, let me know and I can take it into consideration.

---

As an extra I can give you an additional example on how I've currently have my own setup in a Symfony `6.4` application:

<details><summary><code>config/services.yaml</code></summary>
<p>

```yaml
services:
  custom.http_client:
    class: Symfony\Component\HttpClient\Psr18Client

  bunny.client:
    class: ToshY\BunnyNet\Client\BunnyClient
    bind:
      $client: '@custom.http_client'

  bunny.base.api:
    class: ToshY\BunnyNet\BaseAPI
    bind:
      $apiKey: '%env(BUNNY_API_KEY)%'
      $client: '@bunny.client'
```

</p>
</details> 

And then in your service you can use them like so:

<details><summary><code>src/Service/AwesomeBunnyService</code></summary>
<p>

`src/Service/AwesomeBunnyService`
```php

declare(strict_types=1);

namespace App\Service;

use ToshY\BunnyNet\BaseAPI;

class AwesomeBunnyService
{
    public function __construct(
        private readonly BaseAPI $bunnyBaseAPI
    ) {
    }
}
```

</p>
</details> 

---

TLDR;

Provide a HTTP client that implements `Psr\Http\Client\ClientInterface`.

- [`Symfony\Component\HttpClient\Psr18Client`](https://github.com/symfony/http-client/blob/7.0/Psr18Client.php)
- [`GuzzleHttp\Client`](https://github.com/guzzle/guzzle/blob/7.8/src/Client.php)
- ...or any other client implementing `Psr\Http\Client\ClientInterface`

**Comment by tacman** [comment]
Have you considered creating a Symfony bundle around this library?  So you could skip the services.yaml in the application and just let it autowire?

**Comment by ToshY** [comment]
> Have you considered creating a Symfony bundle around this library? So you could skip the services.yaml in the application and just let it autowire?

Yes I have on several occasions.

In earlier days I didn't create a bundle basically because I lacked experience in both PHP as well as Symfony, and I focussed more into finding ways on making this library easier to maintain (for myself) and more flexible (like using PSR standards for example).

While I recently thought about it again I eventually decided not to. The reason: why would I spend time in creating a (framework specific) bundle while you can achieve the same result with a few extra lines in the config 🤔. If I'm not going to use it myself, creating/maintaining one would then feel more like a chore for me than actual fun to do (if that makes sense).

That being said, if you or someone else wants to create a bundle around this library go for it 🙂
